openhab-snap implementation

kubiko opened the pull request.

kubiko committed:
Initial snappy openhab2 commit

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Renaming from openhab2 to openhab

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Cleaning more .bat scripts

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Switching to inline editing

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
clean more unused scripts

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
start using SNAP_COMMON instead of user one

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Thining java runtime installation

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Switching to online openhab distro

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Moving java to the root and ditching rest of stage-package content

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Switching temporarily to devmode

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Temporarily disable daemon, for manual control

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Moving java jre into jvm dir

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Moving java into separate part

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Revert "Temporarily disable daemon, for manual control"

This reverts commit fcc39aa35847bdb1b1d1a31f446a2a51fb9ee2ff.

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Moving to strict confinement

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Adding .gitignore

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Adding icon

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
- moving java back to default place - add process control to console

jdk is using relative paths to access configs in etc/
if we move jdk to different path, those links are broken

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Adding helper script to connect interfaces

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Adding comments to snapcraft.yaml

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Add to description port number for webgui

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Remove service restart condition (always)

Default systemd service option is: restart the service only if it didn't exit 0.
Which is what we want here, so openhab can shut itself down

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
fix: helper scripts are copied twice, remove intance in root of snap

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
make set JAVA_HOME flexible for various java runtimes

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Add build dependency to openjdk-8-jre-headless as build will fail to resolve link to java certificate in clean build environment

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Bumping up version

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
changing grade to 'stable' so we can publish to stable channel

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Create README.md

Adding readme.md

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Updating icon to openHAB one

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Bumping version to 2.0.0_b4-4

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Moving helper scripts to different directory

This makes safer to use dump plugings from parts

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Adding congiguration option for ports

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Add workaround for snapctl get limitations

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
fixing git ignore

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Adding custom java plugin

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Fixing java pluging to work with zulu if available

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Remove JAVA_HOME as it's added by java pluging now

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Cleaning snapcraft.yaml to use custom java plugin

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Cleaning custom plugin implementation

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Fixing wrong PATH to java/bin

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Cleaning custom java plugin

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Fixing link to armhf zulu release

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Renaming jre custom plugin

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Handling default case when setting does not exist

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Fixing configure hook to work properly

snapctl does not yet return all values in the hook, only one which changed, so we need to remember other values

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Editing description, how to change ports

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Clean configure hook

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Bumping up version

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
changing to cleaner configure hook implementation

use of arrays makes this more flexible for future if we need to add
more keys

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
get rid of TODO for SNAP_DATA

There is not really case for use of SNAP_DATA, we should not store
anything service specific to user’s space

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
use bash, as it gives us better array support

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
improve configure hook implementation

cleaning configure hook and making settings handling more elegant and
flexible

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
adding help application

adding help app (openhab.help) with general help
Also adding description how to invoke configure hook
what configure keys are supported and their default values

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
fixing help app definition

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
improving help application

enable interpretation for echo
silence config reading

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
fixing configure hook on all snap system

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kubiko committed:
Tweaking readme

Signed-off-by: Ondrej Kubik <ondrej.kubik@canonical.com>

kaikreuzer commented:
Very cool, thanks @kubiko!

@r2geo I assume you are going to review/merge this PR?
Please let me know, if I can assist in any way.

r2geo commented:
Thanks Ondrej!
A few outstanding questions:
- Do you have and ETA for the dedicated ubuntu core image with specific USB devices enabled? Who should I be in touch with? 
- Are you still planning on a dedicated java snap? 
- Launchpad: how can we best go about publishing openhab-snap. Until now you have done this. Would you like me to take this over? Will that overwrite the snap you created?
  Thanks,
  Roeland

kubiko commented:
Hi
sorry for delay:
- custom pi3 image with serial interfaces for give usb devices:
  http://people.canonical.com/~okubik/snappy/pi3-ondra-20161107-1.img.xz
  it has also openhab preinstalled
  you can see supported USB device here, in case I made mistake
  cat /snap/pi3-ondra/current/meta/snap.yaml 
- I tested dedicated java snap part, but it failed as I needed to write custom plugin to accomplish this, and parts are at the moment limited, and can't have custom plugins. Instead I included this custom plugin directly to openhab, so it's already there. It will use zulu for armhf and amd64, and we can just add links for other arches once zulu starts supporting them, for rest it will use openjdk
- once you create git repo in Launchpad, you should see list of branches in the view
  https://code.launchpad.net/~w-ondra/+git/openhab
  click on branch you want to create snap package for:
  check here branch offline:
  https://code.launchpad.net/~w-ondra/+git/openhab/+ref/offline
  in section "Related snap packages" you will see "Create snap package"
  For my branch you already see there 1 existing snap package. If you click on it, you should see recent builds:
  https://code.launchpad.net/~w-ondra/+snap/openhab-offline
  When you are creating new snap package, pick
  "Ubuntu Xenial, for Ubuntu Core 16"
  git should be picked automatically
  you can also set automatic store upload. That will make it fire and forget. It will auto build/push to store once you push something new to that branch.

kubiko commented:
Kai can you please also create offline branch  in the repository as online version does not work, and I did not have yet time to debug it to see what is the reason for it. It fails at first boot, when trying to install one of the core parts.
I'm at the moment pushing to the store offline version, but once we fix the issue, we can switch to online version and it will auto update to everybody....

r2geo commented:
Thanks a lot Ondrej.

I was messing around trying to create a parallel openhab snap under my launchpad account thinking that you wanted to hand this part over too, but using yours sounds much better for now.

Best regards,

Roeland

> Op 7 november 2016 om 18:13 schreef Ondrej notifications@github.com:
> 
> ```
> Hi
> sorry for delay:
> 
>     * custom pi3 image with serial interfaces for give usb devices:
>       http://people.canonical.com/~okubik/snappy/pi3-ondra-20161107-1.img.xz http://people.canonical.com/%7Eokubik/snappy/pi3-ondra-20161107-1.img.xz
>       it has also openhab preinstalled
>       you can see supported USB device here, in case I made mistake
>       cat /snap/pi3-ondra/current/meta/snap.yaml
> 
>     * I tested dedicated java snap part, but it failed as I needed to write custom plugin to accomplish this, and parts are at the moment limited, and can't have custom plugins. Instead I included this custom plugin directly to openhab, so it's already there. It will use zulu for armhf and amd64, and we can just add links for other arches once zulu starts supporting them, for rest it will use openjdk
> 
>     * once you create git repo in Launchpad, you should see list of branches in the view
>       https://code.launchpad.net/~w-ondra/+git/openhab https://code.launchpad.net/%7Ew-ondra/+git/openhab
>       click on branch you want to create snap package for:
>       check here branch offline:
>       https://code.launchpad.net/~w-ondra/+git/openhab/+ref/offline https://code.launchpad.net/%7Ew-ondra/+git/openhab/+ref/offline
>       in section "Related snap packages" you will see "Create snap package"
>       For my branch you already see there 1 existing snap package. If you click on it, you should see recent builds:
>       https://code.launchpad.net/~w-ondra/+snap/openhab-offline https://code.launchpad.net/%7Ew-ondra/+snap/openhab-offline
>       When you are creating new snap package, pick
>       "Ubuntu Xenial, for Ubuntu Core 16"
>       git should be picked automatically
>       you can also set automatic store upload. That will make it fire and forget. It will auto build/push to store once you push something new to that branch.
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub https://github.com/openhab/openhab-snap/pull/1#issuecomment-258898488 , or mute the thread https://github.com/notifications/unsubscribe-auth/AUoMOICa-X5jymG03n8GvHh8B2MY_DOdks5q71xFgaJpZM4Kl2Sm .
> ```

kaikreuzer commented:
> Kai can you please also create offline branch in the repository as online version does not work

Hi @kubiko, you mean an additional branch here in the git repo? Would you want this temporarily while fixing the online version or shall we have both options on the long run?

kubiko commented:
yeah here in same git repo
Really up to you, I saw that you do offer for download both options, online and offline version of openHAB. So if we are to offer both options as snap, then we would need two branches, also to make auto build lot easier.
In this case we should probably register in the store another name: openhab, and openhab-offline

kubiko commented:
@r2geo plan was always to hand it snap maintenance to openhab community. Using mine git repo or yours does not really make much difference, once ownership is transferred. You set it up only once in launchpad and then you only need to keep pushing to that git repo.
Difference is that if I know that git repo, I can't resist temptation and keep pushing changes directly without 2nd opinion or review :)

kaikreuzer commented:
> In this case we should probably register in the store another name: openhab, and openhab-offline

I think I'd prefer to decide for the one or the other for the snap - having two versions in the store is probably confusing for many.

The decision probably depends on the best upgrade strategy for the application. Can a user decide when to upgrade his snap? Then the offline distro would be just fine; if updates are automatically pushed in the background all the time, it might be more important to reduce the size and go for the online version (which in turn will again download the add-ons that the user has installed).

> plan was always to hand it snap maintenance to openhab community.

I'd suggest that I register an account at Launchpad for the openHAB Foundation as this is the legal entity that publishes openHAB apps in the different stores. I'd then give access to this account to @r2geo.

r2geo commented:
Hi Kai,

Why do you think it would be more confusing than the standard openhab offline and online option? You currently provide both option when people install.

Maybe 2 snaps can be a temporary solution for now? At the moment the snap with openhab offline works, but the online version would be ideal in the long run (I think).

Cheers,

Roeland

> Op 11 november 2016 om 16:26 schreef Kai Kreuzer notifications@github.com:
> 
> ```
>     > > 
> ```
> 
> > ```
> >     In this case we should probably register in the store another name: openhab, and openhab-offline
> > 
> > > 
> > I think I'd prefer to decide for the one or the other for the snap - having two versions in the store is probably confusing for many.
> > ```
> 
> ```
> The decision probably depends on the best upgrade strategy for the application. Can a user decide when to upgrade his snap? Then the offline distro would be just fine; if updates are automatically pushed in the background all the time, it might be more important to reduce the size and go for the online version (which in turn will again download the add-ons that the user has installed).
> 
>     > > 
> ```
> 
> > ```
> >     plan was always to hand it snap maintenance to openhab community.
> > 
> > > 
> > I'd suggest that I register an account at Launchpad for the openHAB Foundation as this is the legal entity that publishes openHAB apps in the different stores. I'd then give access to this account to @r2geo https://github.com/r2geo .
> > ```
> 
> ```
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub https://github.com/openhab/openhab-snap/pull/1#issuecomment-259981307 , or mute the thread https://github.com/notifications/unsubscribe-auth/AUoMOG0KtM9gsoL8csPrinbN20Jvr967ks5q9Ik0gaJpZM4Kl2Sm .
> ```

kaikreuzer commented:
> You currently provide both option when people install.

Well, for the people that set it up themselves and who take care of it themselves.
For me, the major idea of the snaps is to make it very easy for users to install and upgrade; and in other cases of "easy-to-start" packaging like openHABian or the Pine64 SD-card images, we also only provide the offline distro and do not leave a choice - because it does not really bring any benefit to the user.

> but the online version would be ideal in the long run (I think).

Why do you think so?

r2geo commented:
Well, the idea of ubuntu core is to provide a small linux install on top of which users can install what they want. In the same way it would be ideal to install a small openhab core and users only install the bindings needed, to keep a minimum footprint. Was that not the idea you had when moving to OH2?

> Op 11 november 2016 om 18:27 schreef Kai Kreuzer notifications@github.com:
> 
> ```
>     > > 
> ```
> 
> > ```
> >     You currently provide both option when people install.
> > 
> > > 
> > Well, for the people that set it up themselves and who take care of it themselves.
> > For me, the major idea of the snaps is to make it very easy for users to install and upgrade; and in other cases of "easy-to-start" packaging like openHABian or the Pine64 SD-card images, we also only provide the offline distro and do not leave a choice - because it does not really bring any benefit to the user.
> > ```
> 
> ```
>     > > 
> ```
> 
> > ```
> >     but the online version would be ideal in the long run (I think).
> > 
> > > 
> > Why do you think so?
> > ```
> 
> ```
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub https://github.com/openhab/openhab-snap/pull/1#issuecomment-260009513 , or mute the thread https://github.com/notifications/unsubscribe-auth/AUoMOCWXCFa8tY5VBnOaRAmTt9SA_SSfks5q9KWUgaJpZM4Kl2Sm .
> ```

kubiko commented:
@r2geo is it install size what worries you or downloading fat offline snap when getting update?
Because we are moving towards deltas and eventually only delta should be downloaded for new version
Otherwise I agree with @kaikreuzer that having two versions in store in long run would be confusing. Especially as snap should be easy way to run openhab.

kaikreuzer commented:
@r2geo Well, I see your point - the offline distro is increasing day by day and most users only need a fraction of its content. So I'd also be ok to go for the online distro, considering that the upgrade process is anyhow done via the Internet (and not manually for what the offline distro is mainly meant).

kubiko commented:
Finally I had some time to debug open issues and now there is new version of openhab in edge and beta channels. This is still based on beta 4, but this time online version. For illustration online release snap size comes to 80M compare to ~180M of offline version.
Fixed issues:
- online version now starts properly
- fixed functionality for some online bindings like yahoo weather
- console app is now working
- client app is now working
- service is now running as true service, previous version did not pass 'service' to karaf at start
- alternative UIs also seems to be working now as expected

- I redirected user.home location from user's home to snap's home. This is fixed quite few issues. By default snaps can't access hidden files/directories in user's home. Service runs as root, but user can often just call openhab.status or openhab.console as user, this was confusing karaf. Even more with this redirection, now when openhab is uninstalled system is properly cleaned as all writable files should live under /var/snap/openhab/common and system takes care of this.

overall this version seems to be working quite nice
you can install it by calling:
$ snap install —beta openhab

Let me know how this works, and if all good I can promote this version to stable channel, so people get this version as default.

All the changes were also tested with new folder layout built from openhab-distro

Some questions:
- what is the difference between client and console?
- now when we can move with default install to online version, do you want to register openhab-offline version, or what is final decision there?

There is also ongoing effort to enable Launchpad building directly from github repositories, once this lands there won't be need to replicate openhab git repo in launchpad.

kaikreuzer commented:
Sounds great, many thanks!

> what is the difference between client and console?

What exactly do you refer to?

> now when we can move with default install to online version, do you want to register openhab-offline version, or what is final decision there?

I am fine with having only one "openHAB" app, which is the online distro.

> There is also ongoing effort to enable Launchpad building directly from github repositories, once this lands there won't be need to replicate openhab git repo in launchpad.

Sounds good. Do you have any ETA for this? I have today created a Launchpad account on behalf of the openHAB Foundation, which should be the entity publishing the app. Shall I import the current git repo into Launchpad? @r2geo, I can give you access then you could deal with that yourself directly.

kubiko commented:
I can see runtime/bin/client  and runtime/bin/shell
both give me shell to openhab server. What is the difference between two?

No ETA for that feature, sorry I don't have visibility of that, only that it is in works.

So for the moment I guess easiest is to have replicated openhab-snap repo in your launchpad account and keep git pushing there from github repository. We can set up launchpad in a way that it automatically builds every change and pushes snap to the store.
Depends on level of confidence, we can set auto build to auto push to edge/beta channel, and then once happy with testing, you can manually promote revision to the stable channel. Like this people eager to live on the edge can be on edge channel and keep getting latest and greatest.
@r2geo let me know when you are ready with account and repo setup in Launchpad, and I will talk to people in the store, so we can transfer ownership of openhab snap to openHAB Foundation.

for the future we should actually set maven build instead of using url to prebuilt distributions, so we can keep building from tip.

kaikreuzer commented:
> I can see runtime/bin/client and runtime/bin/shell
> both give me shell to openhab server. What is the difference between two?

That's probably a question for Karaf gurus. From the [documentation](https://karaf.apache.org/manual/latest/#_sshd_server), I would assume that "shell" only provides a local, direct shell access, while "client" internally uses ssh and thus only works if ssh is enabled (which is the case for openHAB for local access).
